File: MTWDM iOS Xamarin/App13/App13/ViewController.cs
using System;
using UIKit;
using System.Linq;
using System.Collections.Generic;
using Foundation;

namespace App13
{
    public partial class ViewController : UITableViewController
    {
        Modelo modelo;
        List<string> alfabeto;
        string[] alfabetoarray;

        protected ViewController(IntPtr handle) : base(handle)
        {
            // Note: this .ctor should not contain any initialization logic.
        }

        public override void ViewDidLoad()
        {
            base.ViewDidLoad();
            // Perform any additional setup after loading the view, typically from a nib.

            modelo = new Modelo();

            alfabeto = modelo.lugares.Keys.ToList();
            alfabeto.Sort();
            alfabetoarray = alfabeto.ToArray();
            Console.WriteLine(alfabetoarray);
        }

        public override void DidReceiveMemoryWarning()
        {
            base.DidReceiveMemoryWarning();
            // Release any cached data, images, etc that aren't in use.
        }

       
        public override nint NumberOfSections(UITableView tableView)
        {
            return modelo.lugares.Count;
        }

        public override nint RowsInSection(UITableView tableView, nint section)
        {
            var letra = alfabetoarray[section];

            var countLetras = alfabetoarray.Count();


            switch ((int)section)
            {
                case int n when (n<=countLetras):
                    return modelo.lugares[letra].Count();
                default:
                    return 1;
            }


            /*
             //SWIFT
                 let letra = alfabetoArreglo[section]
        
                    switch section {
                    case 0...alfabetoArreglo.count: return modelo.lugares[letra]!.count
                        //case 0...14: return modelo.lugares[letra]!.count
                    default: return 1
                    }
            */
        }

        public override UITableViewCell GetCell(UITableView tableView, NSIndexPath indexPath)
        {

            var cell = tableView.DequeueReusableCell("lugaresCell", indexPath);

            var letra = alfabetoarray[indexPath.Section];

            var countLetras = alfabetoarray.Count();

            switch (indexPath.Section)
            {
                case int n when (n <= countLetras):
                    var contenido = modelo.lugares[letra];
                    var row = contenido[indexPath.Row];
                    cell.TextLabel.Text = row;
                    break;
                default: break;
            }

            return cell;


            /*
             //SWIFT

                       let cell = tableView.dequeueReusableCell(withIdentifier: "lugaresCell", for: indexPath)
        
                        let letra = alfabetoArreglo[indexPath.section]
                        
                        switch indexPath.section {
                            
                        //case 0...14: cell.textLabel?.text = modelo.lugares[letra]?[indexPath.row]
                        case 0...alfabetoArreglo.count: cell.textLabel?.text = modelo.lugares[letra]?[indexPath.row]
                        default: break
                        }
                        
                        return cell
            */

        }

        public override string TitleForHeader(UITableView tableView, nint section)
        {
            var letra = alfabetoarray[section];

            var countLetras = alfabetoarray.Count();

            switch ((int)section)
            {
                case int n when (n <= countLetras):
                    return letra;
                default: return letra; 
            }
        }
    }
}

File: MTWDM iOS Swift /appPracticaLugares/appPracticaLugares/ViewController.swift
//
//  ViewController.swift
//  appPracticaLugares
//
//  Created by Felipe Hernandez on 06/03/20.
//  Copyright © 2020 Felipe Hernandez. All rights reserved.
//

import UIKit

var listaModelo = [Modelo]()

class ViewController: UITableViewController {

    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view.
    }

    override func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return listaModelo.count
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "lugaresCell", for: indexPath)
        
        cell.textLabel?.text = listaModelo[indexPath.row].nombre
        cell.detailTextLabel?.text = listaModelo[indexPath.row].descripcion
        cell.imageView?.image = listaModelo[indexPath.row].image
        
        return cell
    }
    
    override func viewWillAppear(_ animated: Bool) {
        tableView.reloadData()
    }

}

